library/src/main/java/io/ona/kujaku/location/clients/GoogleLocationClient.java

```diff
@@ -0,0 +1,298 @@
+
+    public GoogleLocationClient(@NonNull Context context) {
+        this.context = context;
+        locationManager = (LocationManager) context.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
+        googleLocationCallback = new GoogleLocationCallback();
+    }
+
+    @Override
+    public void stopLocationUpdates() {
+        if (isMonitoringLocation()) {
+            lastLocation = null;
+            setLocationListener(null);
+            fusedLocationClient.removeLocationUpdates(googleLocationCallback);
+        }
+
+        unregisterForGpsStatusStop();
+    }
+
+    @Override
+    @Nullable
+    public Location getLastLocation() {
+        return lastLocation;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        lastLocation = location;
+
+        if (getLocationListener() != null) {
+            getLocationListener().onLocationChanged(location);
+        }
+    }
+
+    @Override
+    public void requestLocationUpdates(@NonNull android.location.LocationListener locationListener) {
+        LocationRequest locationRequest = new LocationRequest();
+        locationRequest.setInterval(updateInterval);
+        locationRequest.setFastestInterval(fastestUpdateInterval);
+        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+        requestLocationUpdates(locationListener, locationRequest);
+    }
+
+    public void requestLocationUpdates(@NonNull android.location.LocationListener locationListener
+            , @NonNull LocationRequest locationRequest) {
+        setLocationListener(locationListener);
+        if (isProviderEnabled()) {
+            try {
+                fusedLocationClient.getLastLocation()
+                        .addOnSuccessListener((Activity) context, new OnSuccessListener<Location>() {
+                            @Override
+                            public void onSuccess(Location location) {
+                                if (lastLocation == null || location.getTime() > lastLocation.getTime()) {
+                                    lastLocation = location;
+                                }
+                            }
+                        });
+
+                fusedLocationClient.requestLocationUpdates(locationRequest, googleLocationCallback, null);
+
+                // This method protects itself from multiple calls
+                registerForGpsStatusStop();
+            } catch (SecurityException e) {
+                Log.e(TAG, Log.getStackTraceString(e));
+                Toast.makeText(context, R.string.location_disabled_location_permissions_not_granted, Toast.LENGTH_LONG)
+                        .show();
+            }
+        } else {
+            setLocationListener(null);
+            Log.e(TAG, "The provider (" + getProvider() + ") is not enabled");
+        }
+    }
+
+    private void registerForGpsStatusStop() {
```

Comment:
Could this be clearer like `registerForGpsStoppedEvent`, and same for unregister?